Add `slotIdTime :: SlotId -> UTCTime` to WalletLayer

Anviking opened the pull request:
# Issue Number

#465 


# Overview

- [x] Added a naive `slotIdTime` to the `WalletLayer` (agnostic to changing blockchain parameters)
- [x] Moved `waitForWalletSync` to WalletLayer
- [x] Duplicated http-bridge `Cardano.WalletSpec` to Jormungandr, and added a test that the wallet tip, once synced, corresponds to a time in `(now - slotLength, now)`

# Comments

⚠️ also includes #572, but the base branch is still master ⚠️ 

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -357,13 +360,15 @@ newWalletLayer
     -> Block (Tx t)
-        -- ^ Very first block
+        -- ^ Very first block (block0)
+    -> UTCTime
```
See also #527 because this is also adding config parameters to the wallet layer.

rvl commented on lib/http-bridge/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -62,6 +64,22 @@ spec = do
             unless (tip > (SlotId 0 0)) $
                 expectationFailure ("The wallet tip is still " ++ show tip)
+
+        it "The time of the tip is recent" $ \(_, wallet) -> do
```
I'm not sure whether "The time of the tip is recent" is true in general. The network can miss blocks for example, or the node could be temporarily disconnected from its peers.

rvl reviewed (commented):
:+1:

Anviking committed:
Add slotIdTime to WalletLayer

Anviking committed:
Test that the slotIdTime of the wallet tip is recent

paweljakubas commented on exe/wallet/jormungandr/Main.hs:
```diff
@@ -352,7 +352,7 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
                 Left (ErrGetBlockchainParamsGenesisNotFound _) ->
                     handleGenesisNotFound (sb, tracer)
-                Left (ErrGetBlockchainParamsNoInitialPolicy _) ->
+                Left (ErrGetBlockchainParamsIncompleteParams _) ->
```
indeed, better error name

paweljakubas commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -720,6 +732,18 @@ newWalletLayer tracer bp db nw tl = do
+        sl0 = block0 ^. #header ^. #slotId
+
+        convert :: DiffTime -> NominalDiffTime
+        convert = toEnum . fromEnum
```
we can also use the fact that that `NominalDiffTime` and `DiffTime` both have `RealFrac` instances. Hence they have both `Real` and `Fractional` instances. If so we can ,
```
convert :: DiffTime -> NominalDiffTime
convert = fromRational . toRational
```

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -705,7 +705,7 @@ fromFlatSlot n = SlotId e (fromIntegral s)
 
 epochLength :: Integral a => a
-epochLength = 21600
+epochLength = 500
```
why this change?

Anviking replied:
This was temporary. Jormungandr uses `500`, so without this the "slotIdTime of tip is recent"-test on jormungandr didn't pass. So we need to at least use a dynamic `SlotsPerEpoch`/`epochLength` from `BlockchainParameters` 🤔

paweljakubas commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs:
```diff
@@ -158,6 +161,14 @@ block0 = Block
+-- Retrieved from https://github.com/input-output-hk/cardano-sl/blob/403e26a1d33623f4537517535b7de37773ae9259/scripts/prepare-genesis/README.md#31-genesis-json-format
+-- and also matches 'lib/mainnet-genesis.json' in cardano-sl.
+byronBlock0Date :: UTCTime
+byronBlock0Date = posixSecondsToUTCTime 1505621332
```
:+1:

KtorZ replied:
Hmmmm.... This depends on the type of network, doesn't it? Also, this time seems wrong for mainnet anyway :s 
See also: 

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/http-bridge/test/bench/Main.hs#L323-L326

paweljakubas reviewed (approved):
very nice PR, I am fond of tests added! You are very close to nail it :-) (some more places needs revision, you need to import `block0Date` in `lib/core/test/unit/Cardano/WalletSpec.hs` etc)

paweljakubas reviewed (changes requested):
probably some more places will need updating of new `BlockchainParameters` definition:
```
[pawel@arch cardano-wallet]$ ag BlockchainParameters
lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs
41:    ( BlockchainParameters (..) )
176:    , getInitialBlockchainParameters
178:        -> ExceptT ErrGetBlockchainParams m (BlockchainParameters (Jormungandr n))
241:    , getInitialBlockchainParameters = \block0 -> do
270:                return $ BlockchainParameters

lib/jormungandr/test/integration/Main.hs
30:    ( BlockchainParameters (..), newWalletLayer )
207:    -> IO (NetworkLayer (Jormungandr n) IO, BlockchainParameters (Jormungandr n))
214:        getInitialBlockchainParameters jormungandr (coerce block0H)

lib/core/src/Cardano/Wallet.hs
24:    , BlockchainParameters (..)
368:data BlockchainParameters t = BlockchainParameters
383:    -> BlockchainParameters t
410:    BlockchainParameters block0 feePolicy (SlotLength slotLength) txMaxSize = bp

lib/core/test/unit/Cardano/WalletSpec.hs
20:    ( BlockchainParameters (..)
375:    let bp = BlockchainParameters block0 policy slotLength txMaxSize

lib/http-bridge/test/bench/Main.hs
19:    ( BlockchainParameters (..), WalletLayer (..), newWalletLayer )
215:    let bp = BlockchainParameters

lib/http-bridge/test/integration/Cardano/WalletSpec.hs
16:    ( BlockchainParameters (..), WalletLayer (..), newWalletLayer )
84:        let bp = BlockchainParameters block0 byronFeePolicy byronSlotLength byronTxMaxSize

lib/http-bridge/test/integration/Main.hs
23:    ( BlockchainParameters (..), newWalletLayer )
267:            let bp = BlockchainParameters

exe/wallet/http-bridge/Main.hs
60:    ( BlockchainParameters (..), WalletLayer )
285:            -> IO (NetworkLayer t IO, BlockchainParameters t)
290:            let bp = BlockchainParameters

exe/wallet/jormungandr/Main.hs
63:    ( BlockchainParameters (..), WalletLayer )
77:    ( ErrGetBlockchainParams (..), getInitialBlockchainParameters )
340:            -> IO (NetworkLayer t IO, BlockchainParameters t)
349:                runExceptT (getInitialBlockchainParameters jor (coerce block0H)) >>= \case
```

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -720,6 +738,21 @@ newWalletLayer tracer bp db nw tl = do
+        addUTCTime (convert $ slotLength * slots) block0Date
+      where
+        slots = fromIntegral . fromEnum $ (flatSlot' slN) - (flatSlot' sl0)
+        sl0 = block0 ^. #header ^. #slotId
```
I don't think this should live in the `WalletLayer`. This is a pure function that could live in the `Primitive.Types` module.

Anviking replied:
Tbh, I'm thinking it might be best to go for the full solution (with dynamic slotting) straight away.

KtorZ replied:
Seems reasonable. You don't need to make many assumption about how the slot times are determined in order to implement this function. You can expect an index mapping `SlotNo` (or `(Epoch, SlotId)` to a given `(SlotLength, EpochLength)` and make sure to correctly compute the corresponding time. 

I'd even expect this to be so isolated that it could live in its own PR with dedicated testing. No need to do any modification to the wallet layer to implement this.

Anviking replied:
Or no, I changed my mind 😅 .

Storing a list, `[SlottingConfiguration]`,
```haskell
data SlottingConfiguration 
   { validFrom :: SlotId
    , slotLength :: SlotLength
    , epochLength :: EpochLength
    }
```
would be kind of plausible assuming the slotting configuration doesn't change regularly, but for _me_, the same fears as here https://github.com/input-output-hk/cardano-wallet/pull/563#discussion_r304824349 would apply.

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs:
```diff
@@ -169,3 +183,14 @@ byronSlotLength = SlotLength $ secondsToDiffTime 20
+                    , getSlotLength = byronSlotLength
+                    , getTxMaxSize = byronTxMaxSize
+                    , getSlotsPerEpoch = Quantity 21600
+                    }
```
That's a lot of indentation somehow :sweat_smile:

KtorZ commented on lib/jormungandr/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -0,0 +1,139 @@
+                ] (return ())
+                Inherit
+            ]
+        threadDelay 1000000
```
```suggestion
threadDelay (1 * second)
```

KtorZ commented on lib/jormungandr/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -0,0 +1,139 @@
+                (WalletName "My Wallet")
+                (mkSeqState @(Jormungandr 'Testnet) (xprv, mempty) minBound)
+            unsafeRunExceptT $ restoreWallet wallet wid
+            threadDelay 8000000
```
```suggestion
threadDelay (8 * second)
```

KtorZ commented on lib/jormungandr/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -0,0 +1,139 @@
+                unsafeRunExceptT (readWallet wallet wid)
+            let mintedTime = slotIdTime wallet tip
+            t <- getCurrentTime
+            if t > (addUTCTime (-20) mintedTime)
```
This number looks quite arbitrary, what's the rationale here?

Anviking replied:
The maximum delay should be `SlotDuration = 10 s`, so instead making sure the tip was minted within 20 seconds ago is just for some additional marginal.

KtorZ replied:
Would be good to make this explicit. Especially because the slot duration isn't a "global constant" that necessarily has the same value. We know it's `10` because we set it to `10` in the genesis block. The best would be to expose a `getBlockchainParameters` from the wallet layer and use the slot duration provided there.

Also, if the maximum delay is the slot duration, we should check for the slot duration, and not twice of it.

KtorZ commented on lib/jormungandr/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -0,0 +1,144 @@
+            unsafeRunExceptT $ restoreWallet wallet wid
+            threadDelay (30 * second)
+            -- TODO: This wait time is curcial to allow the wallet to sync,
+            -- but it might be unrealistic to do so.
```
Note that you could also wait for the wallet to be synced by polling it every second or so :) 
That should take far less than 30s on Jörmungandr actually.

Anviking replied:
👍 Thanks! --> #572

Anviking committed:
Move waitForWalletSync & add tracer

- Move waitForWallet sync to a helper function in Cardano.Wallet
- Setup a proper tracer in http-bridge:restore benchmark

Anviking committed:
Add getBlockchainParameters to WalletLayer + fixes

Anviking committed:
Add Buildable instance for `Percentage`

Commit:
Regenerate nix

Anviking commented on lib/jormungandr/test/integration/Cardano/WalletSpec.hs:
```diff
@@ -0,0 +1,154 @@
+                expectationFailure ("The wallet tip is still " ++ show tip)
+
+        -- This tests both that the wallet syncs correctly, and that
+        -- 'slotIdTime' computes the time of the wallet tip correctly.
```
So this test is testing two things… We could just get the tip from the `NetworkLayer` and call `slotIdTime` on it. We could get by without any `waitForWalletSync`. On the other hand I thought it was kind of interesting to use the `currentTip`.
